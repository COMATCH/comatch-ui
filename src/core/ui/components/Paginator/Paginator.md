## Paginator

A wrapper around `react-paginate`.

For a complete reference on the available prop types check the documentation
of the `react-paginate` package, [click here](https://github.com/AdeleD/react-paginate).

prop | type | default | required | description
---- | :----: | :-------: | :--------: | -----------
**className** | `String` | `null` | :x: | 

