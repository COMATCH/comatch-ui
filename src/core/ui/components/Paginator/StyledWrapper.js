import React from 'react';
import ReactPaginate from 'react-paginate';
import styled from 'styled-components';
import { palette } from '../../../styles/variables';

export const StyledWrapper = styled.div`
    list-style: none;

    li {
        background-color: ${palette.white};
        border: 1px solid ${palette.midGray};
        color: ${palette.primary};
        display: inline-block;
        transition: color 250ms ease-out, background-color 250ms ease-out;
        user-select: none;

        &.next {
            border-top-right-radius: 3px;
            border-bottom-right-radius: 3px;
        }

        &.previous {
            border-top-left-radius: 3px;
            border-bottom-left-radius: 3px;
        }

        &.next.disabled,
        &.previous.disabled {
            color: ${palette.midGray};
            cursor: not-allowed;
        }

        &.selected,
        &:not(.disabled):hover {
            background-color: ${palette.secondary};
            color: ${palette.white};
            cursor: pointer;
        }

        a {
            display: block;
            width: 35px;
            text-align: center;
            padding: 7.5px 0;
        }
    }

    li ~ li {
        border-left: none;
    }

    li,
    a {
        outline: 0 !important;
    }
`;
