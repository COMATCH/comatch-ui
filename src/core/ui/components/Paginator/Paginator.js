import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import ReactPaginate from 'react-paginate';
import { StyledWrapper as StyledReactPaginate } from './StyledWrapper';

const propTypes = {
    className: PropTypes.string,
};

const defaultProps = {
    className: null,
};

/**
 * A wrapper around `react-paginate`.
 *
 * For a complete reference on the available prop types check the documentation
 * of the `react-paginate` package, [click here](https://github.com/AdeleD/react-paginate).
 */

export const Paginator = ({ className, ...props }) => (
    <StyledReactPaginate className={classNames('Paginator', className)}>
        <ReactPaginate breakLabel="…" previousLabel="«" nextLabel="»" {...props} />
    </StyledReactPaginate>
);

Paginator.propTypes = propTypes;

Paginator.defaultProps = defaultProps;
