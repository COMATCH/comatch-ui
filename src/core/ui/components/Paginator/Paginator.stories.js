import React, { Component } from 'react';
import Highlight from 'react-highlight';
import PropTypes from 'prop-types';
import range from 'lodash/range';
import { storiesOf } from '@storybook/react';
import { Paginator } from './Paginator';

// eslint-disable-next-line react/prop-types
const CommentList = ({ data }) => (
    <div id="project-comments" className="commentList">
        <ul>
            {data.map((comment, index) => (
                <div key={index}>{comment.comment}</div>
            ))}
        </ul>
    </div>
);

export class PaginatorExample extends Component {
    static propTypes = {
        perPage: PropTypes.number.isRequired,
        marginPagesDisplayed: PropTypes.number.isRequired,
        pageRangeDisplayed: PropTypes.number.isRequired,
    };

    state = {
        data: [],
        offset: 0,
    };

    loadCommentsFromServer() {
        const { offset } = this.state;
        const { perPage } = this.props;
        const limit = offset + perPage;
        // Supposed server response
        const data = {
            comments: range(offset, limit).map((index) => ({ comment: `Comment #${index + 1}` })),
            meta: {
                total_count: 100,
                limit: 10,
            },
        };
        // Simulate async request
        setTimeout(() => {
            this.setState({
                data: data.comments,
                pageCount: Math.ceil(data.meta.total_count / data.meta.limit),
            });
        }, 200);
    }

    componentDidMount() {
        this.loadCommentsFromServer();
    }

    handlePageClick = (data) => {
        const { selected } = data;
        const offset = Math.ceil(selected * this.props.perPage);

        this.setState({ offset }, () => {
            this.loadCommentsFromServer();
        });
    };

    render() {
        const { marginPagesDisplayed, pageRangeDisplayed } = this.props;

        return (
            <div className="commentBox">
                <CommentList data={this.state.data} />
                <Paginator
                    marginPagesDisplayed={marginPagesDisplayed}
                    pageRangeDisplayed={pageRangeDisplayed}
                    pageCount={this.state.pageCount}
                    onPageChange={this.handlePageClick}
                />
            </div>
        );
    }
}

storiesOf('Paginator', module)
    .add('with margin pages display and page range set to 2', () => (
        <>
            <PaginatorExample marginPagesDisplayed={2} pageRangeDisplayed={2} perPage={10} />
            <Highlight className="html">
                {'<Paginator' +
                    '\n\tmarginPagesDisplayed={2}' +
                    '\n\tpageRangeDisplayed={2}' +
                    '\n\tpageCount={this.state.pageCount}' +
                    '\n\tonPageChange={this.handlePageClick}' +
                    '\n/>'}
            </Highlight>
        </>
    ))
    .add('with margin pages display and page range set to 4', () => (
        <>
            <PaginatorExample marginPagesDisplayed={4} pageRangeDisplayed={4} perPage={10} />
            <Highlight className="html">
                {'<Paginator' +
                    '\n\tmarginPagesDisplayed={4}' +
                    '\n\tpageRangeDisplayed={4}' +
                    '\n\tpageCount={this.state.pageCount}' +
                    '\n\tonPageChange={this.handlePageClick}' +
                    '\n/>'}
            </Highlight>
        </>
    ));
