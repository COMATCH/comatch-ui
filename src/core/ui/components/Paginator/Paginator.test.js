import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import noop from 'lodash/noop';
import { Paginator } from './Paginator';
import '../../../config/tests/setup';

describe('Paginator tests', () => {
    const OPTIONAL_PROPS = {
        breakLabel: '…',
        previousLabel: '«',
        nextLabel: '»',
        onClick: noop,
        'aria-disabled': 'true',
        className: 'Paginator',
        onKeyPress: noop,
        role: 'button',
    };

    it('should render and assert snapshot', () => {
        const paginator = renderer.create(<Paginator/>).toJSON();
        expect(paginator).toMatchSnapshot();
    });

    it('should render correctly with optional props', () => {
        const props = {
            ...OPTIONAL_PROPS,
        };
        const paginatorComponent = mount(<Paginator {...props}/>);
        expect(paginatorComponent.prop('className')).toEqual(props.className);
        expect(paginatorComponent.prop('aria-disabled')).toEqual(props['aria-disabled']);
        expect(paginatorComponent.prop('onClick')).toEqual(props.onClick);
        expect(paginatorComponent.prop('onKeyPress')).toEqual(props.onKeyPress);
        expect(paginatorComponent.prop('role')).toEqual(props.role);
        expect(paginatorComponent.prop('tabIndex')).toEqual(undefined);
    });

    describe('verifying the HTML structure', () => {
        const props = {
            ...OPTIONAL_PROPS,
        };
        it('should expect nodes/dom elements', () => {
            const sectionHeader = mount(<Paginator {...props} aria-disabled={true}/>);
            expect(sectionHeader).toHaveLength(1);
            expect(sectionHeader.find('div')).toHaveLength(1);
            expect(sectionHeader.find('ul').find('li')).toHaveLength(9);
            expect(sectionHeader.find('ul').find('li').find('a')).toHaveLength(9);
            expect(sectionHeader.find('div').hasClass('Paginator')).toBe(true);

            expect(sectionHeader.find('ul').find('li').at(0).hasClass('previous disabled')).toEqual(true);
            expect(sectionHeader.find('ul').find('li').at(8).hasClass('next')).toBe(true);
            expect(sectionHeader.find('ul').find('li').at(4).hasClass('break')).toBe(true);

            const listItem = sectionHeader.find('ul').find('li').find('a');
            expect(listItem.at(0).props()['aria-disabled']).toEqual('true');
            expect(listItem.at(0).text()).toEqual(props.previousLabel);
            expect(listItem.at(1).props()['aria-label']).toEqual('Page 1 is your current page');
            expect(listItem.at(1).props()['aria-current']).toEqual('page');
            expect(listItem.at(1).text()).toEqual('1');
            expect(listItem.at(2).props()['aria-label']).toEqual('Page 2');
            expect(listItem.at(2).text()).toEqual('2');
            expect(listItem.at(3).props()['aria-label'],).toEqual('Page 3');
            expect(listItem.at(3).text()).toEqual('3');
            expect(listItem.at(4).text()).toEqual(props.breakLabel);
            expect(listItem.at(5).props()['aria-label']).toEqual('Page 8');
            expect(listItem.at(5).text()).toEqual('8');
            expect(listItem.at(6).props()['aria-label']).toEqual('Page 9');
            expect(listItem.at(6).text()).toEqual('9');
            expect(listItem.at(7).props()['aria-label']).toEqual('Page 10');
            expect(listItem.at(7).text()).toEqual('10');
            expect(listItem.at(8).props()['aria-disabled']).toEqual('false');
            expect(listItem.at(8).text()).toEqual(props.nextLabel);
        });
    });
});
