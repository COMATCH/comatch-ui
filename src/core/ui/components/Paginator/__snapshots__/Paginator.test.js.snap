// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Paginator tests should render and assert snapshot 1`] = `
<div
  className="Paginator sc-bdVaJa hSgXFY"
>
  <ul>
    <li
      className="previous disabled"
    >
      <a
        aria-disabled="true"
        onClick={[Function]}
        onKeyPress={[Function]}
        role="button"
        tabIndex="0"
      >
        «
      </a>
    </li>
    <li
      className="selected"
    >
      <a
        aria-current="page"
        aria-label="Page 1 is your current page"
        onClick={[Function]}
        onKeyPress={[Function]}
        role="button"
        tabIndex="0"
      >
        1
      </a>
    </li>
    <li>
      <a
        aria-current={null}
        aria-label="Page 2"
        onClick={[Function]}
        onKeyPress={[Function]}
        role="button"
        tabIndex="0"
      >
        2
      </a>
    </li>
    <li>
      <a
        aria-current={null}
        aria-label="Page 3"
        onClick={[Function]}
        onKeyPress={[Function]}
        role="button"
        tabIndex="0"
      >
        3
      </a>
    </li>
    <li
      className="break"
    >
      <a
        onClick={[Function]}
        onKeyPress={[Function]}
        role="button"
        tabIndex="0"
      >
        …
      </a>
    </li>
    <li>
      <a
        aria-current={null}
        aria-label="Page 8"
        onClick={[Function]}
        onKeyPress={[Function]}
        role="button"
        tabIndex="0"
      >
        8
      </a>
    </li>
    <li>
      <a
        aria-current={null}
        aria-label="Page 9"
        onClick={[Function]}
        onKeyPress={[Function]}
        role="button"
        tabIndex="0"
      >
        9
      </a>
    </li>
    <li>
      <a
        aria-current={null}
        aria-label="Page 10"
        onClick={[Function]}
        onKeyPress={[Function]}
        role="button"
        tabIndex="0"
      >
        10
      </a>
    </li>
    <li
      className="next"
    >
      <a
        aria-disabled="false"
        onClick={[Function]}
        onKeyPress={[Function]}
        role="button"
        tabIndex="0"
      >
        »
      </a>
    </li>
  </ul>
</div>
`;
